crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -0,0 +1,645 @@
+    /// all registered systems that match that label will be evaluated (in insertion order).
+    ///
+    /// To provide explicit label(s), use [`register_system_with_labels`](SystemRegistry::register_system_with_labels).
+    #[inline]
+    pub fn register_system<Params, S: IntoSystem<(), (), Params> + 'static>(
+        &mut self,
+        world: &mut World,
+        system: S,
+    ) {
+        let boxed_system: Box<dyn System<In = (), Out = ()>> =
+            Box::new(IntoSystem::into_system(system));
+        let automatic_system_label: SystemTypeIdLabel<S> = SystemTypeIdLabel::new();
+
+        // This avoids nasty surprising behavior in case systems are registered twice
+        if !self.is_label_registered(automatic_system_label) {
+            self.register_boxed_system_with_labels(
+                world,
+                boxed_system,
+                vec![Box::new(automatic_system_label)],
+            );
+        } else {
+            let type_name = std::any::type_name::<S>();
+            warn!("A system of type {type_name} was registered more than once!");
+        };
+    }
+
+    /// Register system a system with any number of [`SystemLabel`]s.
+    ///
+    /// This allows the system to be run whenever any of its labels are run using [`run_systems_by_label`](SystemRegistry::run_systems_by_label).
+    ///
+    /// # Warning
+    ///
+    /// Unlike the `register_system` method, duplicate systems may be added;
+    /// each copy will be called seperately if they share a label.
+    pub fn register_system_with_labels<
+        Params,
+        S: IntoSystem<(), (), Params> + 'static,
+        LI: IntoIterator<Item = L>,
+        L: SystemLabel,
+    >(
+        &mut self,
+        world: &mut World,
+        system: S,
+        labels: LI,
+    ) {
+        let boxed_system: Box<dyn System<In = (), Out = ()>> =
+            Box::new(IntoSystem::into_system(system));
+
+        let collected_labels = labels
+            .into_iter()
+            .map(|label| {
+                let boxed_label: Box<dyn SystemLabel> = Box::new(label);
+                boxed_label
+            })
+            .collect();
+
+        self.register_boxed_system_with_labels(world, boxed_system, collected_labels);
+    }
+
+    /// A more exacting version of [`register_system_with_labels`](Self::register_system_with_labels).
+    ///
+    /// Returns the index in the vector of systems that this new system is stored at.
+    /// This is only useful for debugging as an external user of this method.
+    ///
+    /// This can be useful when you have a boxed system or boxed labels,
+    /// as the corresponding traits are not implemented for boxed trait objects
+    /// to avoid indefinite nesting.
+    pub fn register_boxed_system_with_labels(
+        &mut self,
+        world: &mut World,
+        mut boxed_system: Box<dyn System<In = (), Out = ()>>,
+        labels: Vec<Box<dyn SystemLabel>>,
```

Comment:
Why accept a `Vec` here instead of an `&[]`?
This would allow `register_system` to provide an Array, instead of needing to allocate a Vec.

Reply:
Good eye! The generic API is a little messy; I'm going to stick to a slice here given how niche this method is.